src/main/java/com/bloxbean/cardano/client/backend/api/helper/impl/UtxoTransactionBuilderImpl.java

```diff
@@ -378,6 +382,82 @@ private void checkAndAddAdditionalUtxosIfMinCostIsNotMet(Map<String, Set<Utxo>>
+                .entrySet()
+                .stream()
+                .map(entry -> {
+                    Value value = entry
+                            .getValue()
+                            .stream()
+                            .map(tx -> {
+                                if (LOVELACE.equals(tx.getUnit())) {
+                                    return Value.builder().coin(tx.getAmount()).multiAssets(Arrays.asList()).build();
+                                } else {
+                                    Tuple<String, String> policyIdAssetName = AssetUtil.getPolicyIdAndAssetName(tx.getUnit());
+                                    Asset asset = new Asset(policyIdAssetName._2, tx.getAmount());
+                                    MultiAsset multiAsset = new MultiAsset(policyIdAssetName._1, Arrays.asList(asset));
+                                    return Value.builder().coin(BigInteger.ZERO).multiAssets(Arrays.asList(multiAsset)).build();
+                                }
+                            })
+                            .reduce(Value.builder().coin(BigInteger.ZERO).build(), Value::plus);
+
+                    // TxOut before min ada adjustment
+                    TransactionOutput draftTxOut = TransactionOutput.builder().address(entry.getKey()._2).value(value).build();
+
+                    // Calculate required minAda
+                    BigInteger minRequiredAda = new MinAdaCalculator(protocolParams).calculateMinAda(draftTxOut);
+
+                    // Get the max between the minAda and what the user wanted to send
+                    BigInteger actualCoin = minRequiredAda.max(draftTxOut.getValue().getCoin());
+
+                    // The final value to send (value is ada + all multi assets)
+                    Value finalValue = Value.builder().coin(actualCoin).multiAssets(draftTxOut.getValue().getMultiAssets()).build();
+
+                    // Sum user's fee (if specified)
+                    BigInteger fees = entry.getValue().stream().map(PaymentTransaction::getFee).reduce(BigInteger.ZERO, BigInteger::add);
+
+                    // Costs
+                    BigInteger existingMiscCost = senderMiscCostMap.getOrDefault(entry.getKey()._1.baseAddress(), BigInteger.ZERO);
+
+                    // Calculating if extra costs are required (diff between minAda and actual ada, and add to costs)
+                    BigInteger additionalCost = actualCoin.subtract(draftTxOut.getValue().getCoin());
+                    existingMiscCost = existingMiscCost.add(additionalCost);
+
+                    BigInteger costs = entry.getValue().stream().map(PaymentTransaction::getFee).reduce(BigInteger.ZERO, BigInteger::add);
+                    existingMiscCost = existingMiscCost.add(costs);
+
+                    // Update costs per (sender) base address
+                    senderMiscCostMap.put(entry.getKey()._1.baseAddress(), existingMiscCost);
```

Comment:
@nemo83, thanks for the optimization.

Are we missing  datumHash in case of transfer to a script address ? So if there is a datumHash  set in PaymentTransaction, it should be added to the output.

I am just thinking what would happen when there are two PaymentTransactions (to the same script address), with two different datum hash. Should we also group PaymentTransaction based on datumHash (sender, receiver, datumHash) ?  

Here's the link to existing code https://github.com/bloxbean/cardano-client-lib/blob/914ed1d16878cfa964b78783ddce4cf6b1e4ff2f/src/main/java/com/bloxbean/cardano/client/backend/api/helper/impl/UtxoTransactionBuilderImpl.java#L415

We can merge this PR and create an issue to track as a separate PR.

Please let me know.

Reply:
Or, ignore grouping in case there is datumhash. Probably a safer option.

Reply:
@nemo83 Is it possible to ignore grouping when datumhash is there and just consider it as a separate output ?  In case of script output, I am not sure if we can add this intelligence. It depends on the  logic in validator script and output becomes non spendable based on the logic.

Probably safe to keep these output separate. 

BTW, I just tried contract IT tests. Those are failing because of missing datumHash. The contract call transaction doesn't find any utxo with datum hash as the previous payment transaction ignores datum hash in output :)

Reply:
Gotcha, it makes sense tests were failing. Of course the DatumHash wasnt set.

I'm gonna propose some changes. I believe grouping can still be safely done.

Reply:
Right I've suggested some changed for the datum hash.

Pleasee check code [here](https://github.com/bloxbean/cardano-client-lib/pull/59/files#r781539951).

Although all the smart contracts tests are now passing, the following is constantly failing, in both master and this branch

`TransactionHelperServiceIT. mintTokenWithNormalAccountKeyHash()`

I will try to take a look at it tomorrow.

Reply:
@nemo83 Please ignore the test failure for now. I think it is due to another check-in. I will fix it in the master after merge.